File: Non-procedural (functional) programming/Solved/antihric/mnozinove_operace.hs
module MnozinoveOp where
import Functional

--mnozinoviny

intersection :: Eq a => [a] -> [a] -> [a]
intersection xs = foldr (\y a -> if elem y xs && (not $ elem y a) then y:a else a) []

union :: Eq a => [a] -> [a] -> [a]
union xs ys = unique (xs ++ ys)

difference :: Eq a => [a] -> [a] -> [a]
difference xs ys = foldr (\x a -> if (not $ elem x ys) && (not $ elem x a) then x:a else a) [] xs

setPushFront :: Eq a => a -> [a] -> [a]
setPushFront x xs | elem x xs   = xs
                  | otherwise   = x:xs

setPushBack :: Eq a => a -> [a] -> [a]
setPushBack x xs | elem x xs   = xs
                 | otherwise   = xs++[x]

setAppend :: Eq a => [a] -> [a] -> [a]
setAppend xs ys = unique (xs ++ ys)

symmetricDifference :: Eq a => [a] -> [a] -> [a]
symmetricDifference xs ys = (difference xs ys) `union` (difference ys xs)

cartesianProduct :: [a] -> [b] -> [(a,b)]
cartesianProduct xs ys = [(x,y) | x <- xs, y <- ys]

File: Non-procedural Programming/Solved/antidvorak/mnozinove_operace.pl
?- consult('functional.pl').

%mnozinoviny

elementToAdd(Xs, Y, Acc, [Y | Acc]) :-
    member(Y, Xs),
    \+ member(Y, Acc), !.

elementToAdd(_, _, Acc, Acc).
intersection(Xs, Ys, Zs) :- foldr(elementToAdd(Xs),Ys,[], Zs).

union(Xs, Ys, Zs) :- setAppend(Xs,Ys,Zs).


isIn(Ys, X, Acc, [X | Acc]) :-
    \+ member(X, Ys),
    \+ member(X, Acc), !.

isIn(_, _, Acc, Acc).

difference(Xs, Ys, Zs) :- foldr(isIn(Ys), Xs, [], Zs).

% Ys^C vuci Xs
complement(Xs, Ys, C) :- difference(Xs, Ys, C).

setPushFront(X, Xs, Xs) :- member(X, Xs), !.
setPushFront(X, Xs, [X | Xs]).

setPushBack(X, Xs, Xs) :- member(X, Xs), !.
setPushBack(X, Xs, Ys) :- append(Xs, [X], Ys).

setAppend(Xs, Ys, Zs) :-
    append(Xs, Ys, Z),
    unique(Z, Zs).

symmetricDifference(Xs, Ys, Zs) :-
    difference(Xs, Ys, A),
    difference(Ys, Xs, B),
    union(A, B, Zs).

k2(X, Y, Acc, [X-Y | Acc]).
k1(Ys, X, Acc, Ret) :- foldr(k2(X), Ys, [], Out), append(Acc, Out,Ret).
cartesian_product(Xs, Ys, Zs) :- foldl(k1(Ys), Xs, [], Zs).